src/ports/postgres/modules/graph/bfs.py_in

```diff
@@ -0,0 +1,479 @@
+        @param directed        Graph will be treated as directed if this boolean flag 
+                               is set to TRUE. Graph is treated as undirected by default.
+        @param grouping_cols   The list of grouping columns.
+
+    [1] https://en.wikipedia.org/wiki/Breadth-first_search
+    """
+
+    with MinWarning("warning"):
+
+        INT_MAX = 2147483647
+
+        params_types = {'src': str, 'dest': str}
+        default_args = {'src': 'src', 'dest': 'dest'}
+        edge_params = extract_keyvalue_params(edge_args,
+                                            params_types,
+                                            default_args)
+
+        # Prepare the input for recording in the summary table
+        if vertex_id is None:
+            v_st= "NULL"
+            vertex_id = "id"
+        else:
+            v_st = vertex_id
+        if edge_args is None:
+            e_st = "NULL"
+        else:
+            e_st = edge_args
+        if max_distance is None:
+            d_st= "NULL"
+            max_distance = INT_MAX
+        else:
+            d_st = max_distance
+        if directed is None:
+            dir_st= "NULL"
+            directed = False
+        else:
+            dir_st = directed
+        if grouping_cols is None:
+            g_st = "NULL"
+            glist = None
+        else:
+            g_st = grouping_cols
+            glist = split_quoted_delimited_str(grouping_cols)
+
+        src = edge_params["src"]
+        dest = edge_params["dest"]
+        
+        distribution = m4_ifdef(<!__POSTGRESQL__!>, <!''!>,
+            <!"DISTRIBUTED BY ({0})".format(vertex_id)!>)
+        local_distribution = m4_ifdef(<!__POSTGRESQL__!>, <!''!>,
+            <!"DISTRIBUTED BY (id)"!>)
+
+        _validate_bfs(vertex_table, vertex_id, edge_table,
+            edge_params, source_vertex, out_table, max_distance, directed, glist)
+
+        # Initialize grouping related variables
+        grp_comma = ""
+        and_grp_null_checks = ""
+
+        if grouping_cols is not None:
+            grp_comma = grouping_cols + ", "
+            and_grp_null_checks = " AND " + _grp_null_checks(glist)
+            
+        # We keep a table of every vertex, the distance to that vertex from source
+        # and the parent in the path to the vertex
+        # This table will be updated throughout the execution.
+        dist_col = "dist"
+        parent_col = "parent"
+        curr_dist_val = 0
+
+        # Creating the output table with the appropriate columns and data types
+        plpy.execute("""
+            CREATE TABLE {out_table} AS ( 
+                SELECT
+                    {grp_comma} 
+                    {src} AS {vertex_id}, 
+                    {curr_dist_val}::INT AS {dist_col},
+                    {src} AS {parent_col}
+                FROM {edge_table} 
+                LIMIT 0
+            ) {distribution}""".format(**locals()))
```

Comment:
We don't have to distribute the output table unless it helps in faster execution of this algorithm itself. The consumers of MADlib functions normally use `order by` to view results, or might even redistribute the output table based on their need.
If this is indeed required for faster retrieval/join of data in the queries that follow, then it might be a good idea to consider `{grouping_cols}` too apart from `{vertex_id}`.